ci: Add conditional CI for the containerd backed image store

rumpl opened the pull request:
- replaces / closes https://github.com/moby/moby/pull/45232

**- What I did**

Added CI for the containerd integration

Any PR that has the label `containerd-integration` will also run tests with the snapshotters

Example PR without the needed label, note the skipped jobs https://github.com/rumpl/moby/pull/138

**- How I did it**

Moved everything from `test.yml` into a reusable workflow in `.test.yml` and used that to run tests with graph drivers and containerd snapshotters. Also added a workflow to test snapshotters in windows.

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,559 @@
+name: Reusable
```
For consistency let's keep workflow name the same as filename and add same comments as windows reusable workflow:

```suggestion
# reusable workflow
name: .test

# TODO: hide reusable workflow from the UI. Tracked in https://github.com/community/community/discussions/12025
```

crazy-max reviewed (commented):
Note that we have currently 98 jobs. With this change we would have 98 + 45 (.test) + 21 (.windows) = 164 which would increase queue drastically for any repos on moby org (such as buildkit).

rumpl commented:
> True, but these new jobs don't run for _all_ PRs, only for those that they are needed

This is true... for now. At one point it would be great to test both graph drivers and containerd snapshotters at the same time (once the CI is green).

neersighted commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,561 @@
+      use_snapshotter:
+        required: true
+        type: boolean
+      graph_driver:
```
```suggestion
      storage_driver:
```

Since the flag is actually `--storage-driver`.

rumpl replied:
:facepalm: changed it!

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,483 @@
+      - validate-dco
+    outputs:
+      matrix: ${{ steps.platforms.outputs.matrix }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Create matrix
+        id: platforms
+        run: |
+          matrix="$(docker buildx bake binary-smoketest --print | jq -cr '.target."binary-smoketest".platforms')"
+          echo "matrix=$matrix" >> $GITHUB_OUTPUT
+      - name: Show matrix
+        run: |
+          echo ${{ steps.platforms.outputs.matrix }}
+
+  smoke:
+    runs-on: ubuntu-20.04
+    needs:
+      - prepare-smoke
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ${{ fromJson(needs.prepare-smoke.outputs.matrix) }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Prepare
+        run: |
+          platform=${{ matrix.platform }}
+          echo "PLATFORM_PAIR=${platform//\//-}" >> $GITHUB_ENV
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Test
+        uses: docker/bake-action@v2
+        with:
+          targets: binary-smoketest
+          set: |
+            *.platform=${{ matrix.platform }}
```
Don't think these two jobs should be in the reusable workflow as it does not depend on the storage driver. You can keep it in "test" one.

rumpl replied:
Moved these in `test`

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,483 @@
+
+  docker-py:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 120
+    needs:
+      - build-dev
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Set up runner
+        uses: ./.github/actions/setup-runner
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Build dev image
+        uses: docker/bake-action@v2
+        with:
+          targets: dev
+          set: |
+            dev.cache-from=type=gha,scope=dev
+      - name: Test
+        run: |
+          make -o build test-docker-py
+      - name: Prepare reports
+        if: always()
+        run: |
+          mkdir -p bundles /tmp/reports
+          find bundles -path '*/root/*overlay2' -prune -o -type f \( -name '*-report.json' -o -name '*.log' -o -name '*.out' -o -name '*.prof' -o -name '*-report.xml' \) -print | xargs sudo tar -czf /tmp/reports.tar.gz
+          tar -xzf /tmp/reports.tar.gz -C /tmp/reports
+          sudo chown -R $(id -u):$(id -g) /tmp/reports
+          tree -nh /tmp/reports
+      - name: Test daemon logs
+        if: always()
+        run: |
+          cat bundles/test-docker-py/docker.log
+      - name: Upload reports
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: docker-py-reports
+          path: /tmp/reports/*
```
Same for this one I think

rumpl replied:
We do want to test docker-py with a containerd-backed daemon I think

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,483 @@
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Create matrix
+        id: scripts
+        run: |
+          scripts=$(cd ./hack/validate && jq -nc '$ARGS.positional - ["all", "default", "dco"] | map(select(test("[.]")|not)) + ["generate-files"]' --args *)
+          echo "matrix=$scripts" >> $GITHUB_OUTPUT
+      - name: Show matrix
+        run: |
+          echo ${{ steps.scripts.outputs.matrix }}
+
+  validate:
+    runs-on: ubuntu-20.04
+    timeout-minutes: 120
+    needs:
+      - validate-prepare
+      - build-dev
+    strategy:
+      fail-fast: true
+      matrix:
+        script: ${{ fromJson(needs.validate-prepare.outputs.matrix) }}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Set up runner
+        uses: ./.github/actions/setup-runner
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Build dev image
+        uses: docker/bake-action@v2
+        with:
+          targets: dev
+          set: |
+            dev.cache-from=type=gha,scope=dev
+      - name: Validate
+        run: |
+          make -o build validate-${{ matrix.script }}
```
I think we could move validate jobs to a dedicated "validate" workflow like we do in buildkit: https://github.com/moby/buildkit/blob/master/.github/workflows/validate.yml

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,483 @@
+jobs:
+  validate-dco:
+    uses: ./.github/workflows/.dco.yml
+  build-dev:
+    runs-on: ubuntu-20.04
+    needs:
+      - validate-dco
+    strategy:
+      fail-fast: false
+      matrix:
+        mode:
+          - ""
+          - systemd
+    steps:
+      - name: Prepare
+        run: |
+          if [ "${{ matrix.mode }}" = "systemd" ]; then
+            echo "SYSTEMD=true" >> $GITHUB_ENV
+          fi
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Build dev image
+        uses: docker/bake-action@v2
+        with:
+          targets: dev
+          set: |
+            *.cache-from=type=gha,scope=dev${{ matrix.mode }}
+            *.cache-to=type=gha,scope=dev${{ matrix.mode }},mode=max
+            *.output=type=cacheonly
```
I just realized that `build-dev` will run twice if snapshotter is enabled and therefore caching of the dev-image will be racy. Also this job doesn't depend on the reusable workflow inputs so should not be in this workflow but I don't think a workflow can depend on another one unfortunately.

I think we might need to merge test and snapshotter-test workflows and make them depend on a build-dev job. I think it would work.

rumpl replied:
Right, switched things a bit, let's see if it works :)

TBBle commented:
I don't think it's actually testing the containerd image store.

When I tried something similar (see https://github.com/moby/moby/compare/master...TBBle:moby:hack/containerd_image_store%5E) I found that the env-var `TEST_INTEGRATION_USE_SNAPSHOTTER` didn't affect the tests as desired because we register the Docker daemon as a service. So I had to set registry keys to affect the daemon's environment. [It was pretty easy to do with `reg` between registering the daemon as a service and starting that service](https://github.com/moby/moby/compare/master...TBBle:moby:hack/containerd_image_store%5E#diff-90e64dd59d8246305776c16fc6a5a006ef9458e5d0f7a6d64147f1c9ad97ddedR352-R354).

Anyway, you should see `Enabling containerd snapshotter through the $TEST_INTEGRATION_USE_SNAPSHOTTER environment variable. This should only be used for testing.` in the daemon logs if it's testing the containerd image store ([like this](https://github.com/TBBle/moby/actions/runs/5268221853/job/14259050326#step:21:68)), and I don't see that in, e.g., https://github.com/moby/moby/actions/runs/6109262248/job/16580260658?pr=46402.

Another sign it's not using the snapshotter is that it should have failed all of the tests [during setup](https://github.com/TBBle/moby/actions/runs/5268221853/job/14259050326) due to (I think) the wildly incorrect WCOW mount code in containerd 1.6 (which I contributed, AFAIR) vendored into Docker. That was [replaced with good code in containerd 1.7.2](https://github.com/containerd/containerd/releases/tag/v1.7.2), so I think getting anything meaningful out of the testsuite will depend on #45966.

You can also see a couple of quick-hack code changes I had to get the daemon even start, to deal with two issues: One was that when the containerd storage backend is enabled, the network setup code for builder-next assumes we're on Linux and hence uses an explicit network mode instead of "auto". The other is that the graph driver name (also used to select snapshotter) is different between Docker (`windowsfilter`) and Containerd (`windows`). Windows doesn't honour either the env-var or config option to control the graphdriver name, so you can't override the latter one from the CI script.

I also added the matrix in a different spot, but that should not affect any of the above.

rumpl commented:
> Another sign it's not using the snapshotter is that it should have failed all of the tests [during setup](https://github.com/TBBle/moby/actions/runs/5268221853/job/14259050326) due to (I think) the wildly incorrect WCOW mount code in containerd 1.6 (which I contributed, AFAIR) vendored into Docker.

I did at one point manage to make it build/run and run a container:

![image](https://github.com/moby/moby/assets/99933/725c3cac-5497-4b2c-b8ef-91c1ad19efbc)

But I never pushed the code because it was a quick hack to see what's needed

> Side thought: Are we actually running the Linux build through CI with `TEST_INTEGRATION_USE_SNAPSHOTTER ` anywhere? I don't see in-tree configs, e.g., the Jenkinsfile, mentioning that env-var. I'm assuming containerd snapshotter support is not on-by-default yet; Jenkins doesn't appear to capture the daemon logs, but I see a couple of SKIPs due to `!testEnv.UsingSnapshotter()` so I assume the main Jenkins run is testing graphdrivers.

For Linux we have a [draft PR](https://github.com/moby/moby/pull/45232) that @vvoland opened and rebases from time to time, which is not ideal. That's why I opened this PR. I didn't update Jenkins because we don't have the plugins needed to filter with PR labels as I do in this PR. So basically, we don't _yet_ test on linux but this PR changes that.

> I don't think it's actually testing the containerd image store.

Thanks, let me double check that!

TBBle commented:
The logs from the relevant run have expired, but the problem I was alluding to is triggered by the Docker integration test suite setup code, not usual container-running operations. I assume `COPY` in the classic builder relies on being able to mount the container image locally, and so that should show the same failure. Apart from host-mounting, I think all the relevant stuff landed in the 1.6 series, so running containers and Dockerfiles without `COPY` and similar should work approximately as-is. (That's what I'd hoped to see in my own tests, but the testsuite setup failing denied me that pleasure.)

rumpl commented:
I have rebased this PR and removed the windows part, at least with this we will have CI run on linux/amd64

TBBle commented on .github/workflows/windows-2019.yml:
```diff
@@ -13,9 +13,16 @@ jobs:
       - validate-dco
+    uses: ./.github/workflows/.windows-build.yml
+    with:
+      os: windows-2022
```
Is this supposed to be `windows-2019`?

TBBle commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,427 @@
+on:
+  workflow_call:
+    inputs:
+      use_snapshotter:
+        required: true
+        type: boolean
```
Will calling the workflow twice with different inputs see any conflicts in the testsuite artifacts? I assume GitHub's own artifacts are fine as they're separate jobs, but does for example the codecov upload need to distinguish them? (I didn't see any other artifact uploads after a quick look).

Or should it distinguish them? I guess technically we want total coverage for the whole testsuite run, including both graphdrivers and snapshotters. Although the snapshotter specific coverage will be zero in usual test-runs, so maybe codecov upload should be disabled for `use_snapshotter=true` so we don't get spurious falls in code-coverage on merge-to-main runs, if that's a concern.

I note that Windows has a `send_coverage` input which is only used on Windows 2022, so I assume we want to skip-non-primary test-runs, so my guess is that the best course is to skip the `codecov/codecov-action` steps if `use_snapshotter` is true.

rumpl replied:
I don't see conflicts in the artifacts, the test runs reports are acurate for example.

Gotta check codecov though

rumpl replied:
I think it's fine sending the coverage in both cases, windows has the `send_coverage` because it's testing the same thing on two different windows versions. Whereas with the containerd snapshotter we are testing a completely different code path

rumpl replied:
Reports are no longer broken :)

TBBle commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,427 @@
+      use_snapshotter:
+        required: true
+        type: boolean
+      storage_driver:
+        required: true
+        type: string
```
Do you see general value in making this an input, rather than simply choosing `overlay2` or `overlayfs` for `DOCKER_GRAPHDRIVER` based on `use_snapshotter`?

rumpl replied:
No real value no, you're right, I'll remove the storage driver input

TBBle reviewed (commented):
Comments, but nothing majorly blocking as far as I can see, just some possible opportunities to reduce future work a little bit (and a typo).

My only real concern here is the artifact/code-cov question. The other questions/ideas are Windows-specific and since Windows isn't being enabled here, they can be punted to whoever does that, if you prefer.

TBBle commented on .github/workflows/windows-2019.yml:
```diff
@@ -13,9 +13,16 @@ jobs:
       - validate-dco
+    uses: ./.github/workflows/.windows-build.yml
+    with:
+      os: windows-2022
```
Ah, this was fixed while I was creating the review, but I can't make it resolved myself, apparently.

rumpl replied:
Ha yeah I received the email and fixed that immediately, thanks :)

rumpl commented on .github/workflows/windows-2022.yml:
```diff
@@ -8,18 +8,37 @@ on:
       send_coverage: true
+
+  # Disabled for now until we make containerd snapshotters
+  # work on windows, see https://github.com/moby/moby/pull/46402s
```
Thaks

rumpl commented on .github/workflows/.windows.yml:
```diff
@@ -13,6 +13,10 @@ on:
         required: false
         type: boolean
         default: false
+      use_snapshotter: # always false for now until we update containerd
+        required: false
+        type: boolean
+        default: false
```
Also, we can do this in a followup since we are not calling this with containerd snapshotters enabled any more

TBBle replied:
I just looked more-closely at your link to BuildKit, and wow, today I learned that GitHub Actions lets you do _crimes_.

But yeah, should be feasible to mess with the matrix to insert an appropriate `exclude` clause based on the workflow inputs.

rumpl commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,424 @@
+        name: Download reports
+        uses: actions/download-artifact@v3
+        with:
+          name: i${{ inputs.use_snapshotter && 'snapshotter-' || ''}}ntegration-cli-reports
```
Oh man, I'm so lucky you are here :)

vvoland commented on .github/workflows/.test.yml:
```diff
@@ -0,0 +1,363 @@
+        if: always()
+        uses: actions/upload-artifact@v3
+        with:
+          name: docker-py-reports
```
Should we also give this one a unique name?
```suggestion
          name: ${{ inputs.use_snapshotter && 'snapshotter-' || ''}}docker-py-reports
```

crazy-max replied:
Yes I think we should otherwise it will be mixed with non-snapshotter job

rumpl replied:
Done

TBBle commented:
For the record, I was curious how the test suite would go on Windows now that containerd 1.7.6 has been vendored, so I rebased my version of this PR onto the mainline branch.

Sadly, it still fails [at the first action during busybox build (`RUN mkdir C:\tmp && mkdir C:\bin`)](https://github.com/TBBle/moby/actions/runs/6281375365/job/17059820574), and the logs aren't super-helpful there (`unexpected EOF`), so not sure what went wrong. There's like a 60% chance it has something to do with WCOW stdio though, since that's the main problem source for all of the Docker/Windows/Containerd work. ^_^

<details>

<summary>Merged logs of failed step</summary>

```
2023-09-23T04:13:57.9142767Z MOBY [Information] image pulled [traceID=20a7c6128d72d420a4d28266b1803fd3 spanID=9d94b34734d8a189 digest=sha256:d227d7ab11f4aa0da7779adc31616924cf0b15846a10313d83a644541208f80b remote=mcr.microsoft.com/windows/servercore:ltsc2022]
2023-09-23T04:13:57.915340600Z CTRD level=debug msg="delete image" name="moby-dangling@sha256:d227d7ab11f4aa0da7779adc31616924cf0b15846a10313d83a644541208f80b"
2023-09-23T04:13:57.9285776Z Digest: sha256:d227d7ab11f4aa0da7779adc31616924cf0b15846a10313d83a644541208f80b
2023-09-23T04:13:57.9287309Z Status: Downloaded newer image for mcr.microsoft.com/windows/servercore:ltsc2022
2023-09-23T04:13:57.9563826Z MOBY [Information] debug: [BUILDER] Cache miss: [cmd /S /C mkdir C:\tmp && mkdir C:\bin]
2023-09-23T04:13:57.9563826Z MOBY [Information] debug: [BUILDER] Command to be executed: [cmd /S /C mkdir C:\tmp && mkdir C:\bin] [traceID=20a7c6128d72d420a4d28266b1803fd3 spanID=9d94b34734d8a189]
2023-09-23T04:13:57.9654515Z  ---> d227d7ab11f4
2023-09-23T04:13:57.9654911Z Step 6/13 : RUN mkdir C:\tmp && mkdir C:\bin
2023-09-23T04:13:57.973779100Z CTRD level=debug msg="stat snapshot" key="sha256:313b663315113261ec8ee4b9b3652b3504b995aa3fd9d3f63fdebf2316a3cfe0"
2023-09-23T04:13:57.985329600Z CTRD level=debug msg="prepare snapshot" key=d63a977fa40b8866ec74a25901c5b5d85bc669de6eaea25d398cd4d8b7943ad9-init-key parent="sha256:313b663315113261ec8ee4b9b3652b3504b995aa3fd9d3f63fdebf2316a3cfe0"
2023-09-23T04:13:57.992154900Z CTRD level=debug msg=createSnapshot
2023-09-23T04:13:58.011612200Z CTRD level=debug msg="event published" ns=moby topic=/snapshot/prepare type=containerd.events.SnapshotPrepare
2023-09-23T04:13:58.0955145Z MOBY [Information] debug: Calling proc (1)
2023-09-23T04:13:58.0962106Z MOBY [Information] debug: Calling proc (2)
2023-09-23T04:13:58.186150800Z CTRD level=debug msg="garbage collected" d=8.4145ms
2023-09-23T04:13:58.1874698Z unexpected EOF
2023-09-23T04:13:58.7239645Z ##[error]Process completed with exit code 1.
2023-09-23T04:13:58.7583587Z ##[group]Run $ErrorActionPreference = "SilentlyContinue"
```
</details>

But at least it's pretty easy to get it into a "testable" state, as the only CI things in my branch that this PR doesn't have is the TODO-marked Windows enablement.

I might rebase my branch onto this PR to verify that, and provide a base for further exploration.

thaJeztah commented:
For future reference / future self; we discussed this PR in Yesterday's maintainers call;

- currently, this PR requires the `containerd-integration` label to be set on PRs for it to run
- ☝️ (probably) adding the label _after_ the PR was opened won't trigger it "after-the-fact", but (we can give it some testing once merged) a "restart CI" or good'ol open-and-close should fix that
- we also wanted to update this PR, and have it run against master / main: we _know_ it's "red" (for now), but having CI run on all commits merged into master makes it more visible to track "what's left" on the containerd-integration work
- _BUT_ we also discussed that we should add back some CI -status badges in the README, and we don't want `master` to always be red, so I asked @rumpl to move this CI to a separate file, so that it can get its own badge (separate from "non-containerd-intregration")

I should also add that this is a large diff, and "lotsa YAML", so I'll give it another look, but in doing so will somewhat depend on prior reviews of those that already happened, and I'll ask @crazy-max (as our in-house GHA guru) to give it a final "once-over".

thaJeztah commented:
@vvoland FYI; I added "closes" https://github.com/moby/moby/pull/45232 (as I think we can close that one once this is merged)

rumpl committed:
CI: move smoke, validate, test and build-dev jobs

Moving these jobs to reusable workflows to declutter the test.yaml
workflow, this will also help reuse these workflows to run tests with
the container-snapshotter feature enabled

Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>

rumpl committed:
CI: Add the test-snapshotter workflow

This workflow will run all the tests (unit, integration and
integration-cli) with the containerd-snapshotter feature enalbed.

We only run these tests for pull requests that have the
"containerd-integration" label and also on any other branch/tag.

Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>

TBBle commented on .github/workflows/snapshotter-test.yml:
```diff
@@ -0,0 +1,32 @@
+name: snapshotter-test
```
Since it's an analogue with `test` but using the snapshotter, should it be `test-snapshotter`? Then it should sort adjacent-to but below the `test` workflow, I hope.

This occurred to me when I was thinking about whether we'd want to do split out the snapshotter test job for `windows-2022` the same way, and if so, what to call it.

Also, thinking about future naming if and when the snapshotter becomes the default, i.e. in `test`, what will the non-snapshotter test be named? `non-snapshotter-test` is very English-like, but sorts weirdly, and the most important aspect (`test`) is way over at the end.

rumpl replied:
We could call them `test-graphdriver` and `test-snapshotter`

thaJeztah replied:
> We could call them `test-graphdriver` and `test-snapshotter`

Works for me 👍 

(GitHub grouping / listing checks remains horrible in either case 😂)

rumpl replied:
Guess it's too much to ask to have this sorter alphabetically 
<img width="312" alt="Screenshot 2023-09-29 at 18 13 38" src="https://github.com/moby/moby/assets/99933/6d32a938-6a7b-4a4a-9aae-e0d3e0cfe23b">

thaJeztah commented:
Thanks for splitting! Gonna have a look; one quick thing if you're updating/pushing; there's a typo in the first commit (`s/vlaidate/validate/`);

```
CI: move smoke, vlaidate, test and build-dev jobs
```

rumpl commented:
> Thanks for splitting! Gonna have a look; one quick thing if you're updating/pushing; there's a typo in the first commit (`s/vlaidate/validate/`);
> 
> ```
> CI: move smoke, vlaidate, test and build-dev jobs
> ```

Fixed that typo and another one in the second commit that you didn't see :)

TBBle commented:
Passing thought: Are there unit tests affected by the `TEST_INTEGRATION_USE_SNAPSHOTTER` or `DOCKER_GRAPHDRIVER` env-vars? If not, can we skip them in snapshotter-test? If so, should they be, or is that a future task to refactor so-affected unit tests to always cover both variations as long as they're supported, disirregardless of the environment they run under?

thaJeztah commented:
> Passing thought: Are there unit tests affected by the TEST_INTEGRATION_USE_SNAPSHOTTER or DOCKER_GRAPHDRIVER env-vars? If not, can we skip them in snapshotter-test? If so, should they be, or is that a future task to refactor so-affected unit tests to always cover both variations as long as they're supported, disirregardless of the environment they run under?

I don't think the unit-tests should be affected by this, but (honestly) can't say with 100% certainty (I know that some unit-tests are _not-as-unit-as-they-should_ be (looking at you: "networking unit-tests that depend on iptables").

I definitely think we need to look at the matrices as a whole, as I think there's some unneeded duplication (as well as looking if we can make them more efficient; some checks depend "too much" on others, which means having to wait for others to complete before being able to restart them, whereas others are fast to run (but relatively high overhead for building the initial parts).

All that said; I think it's ok to look at that in  follow-ups (but we could consider creating a tracking ticket for this).


(Also trying to give @rumpl some sleep after waking up in the middle of night, dreaming of YAML and more YAML) 😂

rumpl commented:
Sooo much yaml! 😭😂

But you are right @TBBle, we might be able to run unit tests only once. I’m off for Dockercon next week so won’t have time to take a closer look, so maybe we should review this as is and we can optimize later? The double unit tests only hit those that open a PR in the snapshotter case; ok on master too but no one really waits for that I guess. 

Hold off merging this though, now I kinda want to take a look at this this weekend while on the plane. 

On top of this there are other optimisations we can do to our CI but I would say they are out of scope of what I want to accomplish here but will definitely get back to it while the ~pain~ memory is still fresh

crazy-max commented on .github/workflows/.smoke.yml:
```diff
@@ -0,0 +1,58 @@
+# reusable workflow
+name: .smoke
```
Why do we need a reusable workflow for smoke tests? It seems only called from `test-graphdriver`.

crazy-max replied:
Maybe instead we could just move this to a regular workflow named `test-smoke.yml` as it's not tied to graphdriver or snapshotter.

TBBle replied:
I assume the idea is that if `smoke` fails, we can abort the run early. That's what the name suggests to me, anyway. So perhaps it should be called from both test-runs?

crazy-max commented on .github/workflows/.test.yml:
```diff
@@ -1,131 +1,38 @@
       -
         name: Set up runner
         uses: ./.github/actions/setup-runner
+      -
+        name: Set up tracing
+        uses: ./.github/actions/setup-tracing
```
Don't think we need to set up tracing for unit tests? cc @cpuguy83

cpuguy83 replied:
Correct we aren't doing any tracing in unit tests.

crazy-max commented on .github/workflows/.windows-build.yml:
```diff
@@ -0,0 +1,100 @@
+# reusable workflow
+name: .windows-build
```
Why can't we just keep this in `.windows.yml`? We already have a bunch of workflows, would be nice to avoid unnecessary split.

crazy-max commented on .github/workflows/.windows.yml:
```diff
@@ -249,7 +174,6 @@ jobs:
     runs-on: ${{ inputs.os }}
     timeout-minutes: 120
     needs:
-      - build
```
Related to previous comment, this is racy as we could call this reusable workflow without having a dependency on `build` so I think it's best to keep it `build` here.

crazy-max commented on .github/workflows/.test.yml:
nit: I think it should be a rename of `test.yml` > `test-graphdriver.yml` instead

TBBle replied:
AFAIK in Git, renames are detected automatically at `diff` time, and target minimising diff size. In this case, the bulk of what was in test.yml ended up in .test.yml so it picked that as a rename.

crazy-max reviewed (commented):
I think we could merge `test-graphdriver` and `test-snapshotter` together and create a dynamic matrix if snapshotter is enabled. That would reduce changes footprint significantly but needs to make sure it's possible. I'm looking at it on another branch and keep you posted.

crazy-max commented:
@rumpl Opened https://github.com/moby/moby/pull/46572 as an alternative. Let me know what you think.

crazy-max commented on .github/workflows/test-snapshotter.yml:
```diff
@@ -0,0 +1,32 @@
+  test:
+    if: contains(github.event.pull_request.labels.*.name, 'containerd-integration') || github.event_name != 'pull_request'
+    needs:
+      - build-dev
```
`build-dev` can be called from `test-graphdriver` and `test-snapshotter` which would make the cache racy.

crazy-max commented on .github/workflows/test-snapshotter.yml:
```diff
@@ -0,0 +1,32 @@
+  validate-dco:
+    uses: ./.github/workflows/.dco.yml
+
+  build-dev:
+    needs:
+      - validate-dco
+    uses: ./.github/workflows/.build-dev.yml
```
`build-dev` is called from `test-graphdriver` and `test-snapshotter` which would make the cache racy.

rumpl commented:
Closing in favor of https://github.com/moby/moby/pull/46572
